felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -418,21 +463,87 @@ func newBPFEndpointManager(
 		}
 	}
 
+	// If not running in test
+	if m.dp == m {
+		// Repin jump maps to a differnt path so that exesting programs keep working
+		// as if nothing has changed. We keep those maps as long as we have dirty
+		// devices.
+		//
+		// Since we are restarting, we reload programs for all the devices, the
+		// generic sets and the preables and then we can just remove the old maps.
+		// We never copy from the old maps to the new ones.
+		if err := m.repinJumpMaps(); err != nil {
+			return nil, err
+		}
+		m.removeOldJumps = true
+	}
+
 	return m, nil
 }
 
+func (m *bpfEndpointManager) repinJumpMaps() error {
+	oldBase := path.Join(bpfdefs.GlobalPinDir, "old_jumps")
+	err := os.Mkdir(oldBase, 0700)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("cannot create %s: %w", oldBase, err)
+	}
+
+	tmp, err := os.MkdirTemp(oldBase, "")
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("cannot create temp dir in %s: %w", oldBase, err)
+	}
+
+	mps := []maps.Map{
+		m.bpfmaps.ProgramsMap,
+		m.bpfmaps.PolicyMap,
+		m.bpfmaps.XDPProgramsMap,
+		m.bpfmaps.XDPPolicyMap,
+	}
+
+	for _, mp := range mps {
+		pin := path.Join(tmp, mp.GetName())
+		if err := libbpf.ObjPin(int(mp.MapFD()), pin); err != nil {
+			_ = os.RemoveAll(tmp)
+			return fmt.Errorf("failed to repin %s to %s: %w", mp.GetName(), pin, err)
+		}
+	}
+
+	for _, mp := range mps {
+		if err := mp.Close(); err != nil {
+			_ = os.RemoveAll(tmp)
+			return fmt.Errorf("failed to close %s: %w", mp.Path(), err)
+		}
+		if err := os.Remove(mp.Path()); err != nil {
+			_ = os.RemoveAll(tmp)
+			return fmt.Errorf("failed to remove %s from %s: %w", mp.GetName(), mp.Path(), err)
+		}
+	}
+
+	for _, mp := range mps {
+		if err := mp.EnsureExists(); err != nil {
+			_ = os.RemoveAll(tmp)
+			return fmt.Errorf("failed to recreate %s: %w", mp.Path(), err)
+		}
+	}
+
+	return nil
+}
+
 // withIface handles the bookkeeping for working with a particular bpfInterface value.  It
 // * creates the value if needed
 // * calls the giving callback with the value so it can be edited
 // * if the bpfInterface's info field changes, it marks it as dirty
 // * if the bpfInterface is now empty (no info or state), it cleans it up.
 func (m *bpfEndpointManager) withIface(ifaceName string, fn func(iface *bpfInterface) (forceDirty bool)) {
-	iface := m.nameToIface[ifaceName]
+	iface, ok := m.nameToIface[ifaceName]
+	if !ok {
+		iface = zeroIface
+		iface.dpState.clearPolicies()
+	}
 	ifaceCopy := iface
 	dirty := fn(&iface)
 	logCtx := log.WithField("name", ifaceName)
 
-	var zeroIface bpfInterface
 	if reflect.DeepEqual(iface, zeroIface) {
```

Comment:
I think this test will always be false now so cleanup will never happen.

Reply:
oopps, good catch! Initialized properly now